### PR TITLE
Only avoid duplicate payments if we didn't fail sending (and expose `list_payments`)

### PR DESCRIPTION
**Pull request by tnull**
Previously, we went from not tracking payments when they immediately failed to send to always tracking them but denying any retries for payments whose payment hash was known to us. However, this is of course much too restrictive as payments might fail due to local connections issues or unavailability of usable channels, in which case the invoices/payment hashes would be 'burnt' and couldn't get retried.

Here, we introduce a new `PaymentStatus::SendingFailed` to differentiate send failures and allow retrying tracked payments if they failed to send immediately. We also rename some error types accordingly for further clarification.

Additionally, we expose a simple `Node::list_payments` variant, mainly to be used in bindings where we can't make use of the `_with_filter` variant.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -86,7 +86,9 @@ impl fmt::Display for Error {
 			Self::InvalidInvoice => write!(f, "The given invoice is invalid."),
 			Self::InvalidChannelId => write!(f, "The given channel ID is invalid."),
 			Self::InvalidNetwork => write!(f, "The given network is invalid."),
-			Self::DuplicatePayment => write!(f, "A payment with the given hash has already been initiated."),
+			Self::DuplicatePayment => {
+				write!(f, "A payment with the given hash has already been initiated.")
+			}
```
Comment (jkczyz) [inline src/error.rs]: Looks like this belongs in the previous commit.

File: src/lib.rs [hunk 2]
```diff
@@ -1339,11 +1341,16 @@ impl Node {
...
 				};
 				self.payment_store.insert(payment)?;
 
-				Err(Error::PaymentFailed)
+				match e {
+					channelmanager::RetryableSendFailure::DuplicatePayment => {
+						Err(Error::DuplicatePayment)
+					}
+					_ => Err(Error::PaymentSendingFailed),
+				}
```
Comment (jkczyz) [inline src/lib.rs:1353]: Wouldn't we want to not update the `payment_store` if it is a duplicate?

File: src/lib.rs [hunk 3]
```diff
@@ -1334,22 +1334,23 @@ impl Node {
...
+						let payment = PaymentDetails {
+							preimage: None,
+							hash: payment_hash,
+							secret: payment_secret,
+							amount_msat: invoice.amount_milli_satoshis(),
+							direction: PaymentDirection::Outbound,
+							status: PaymentStatus::SendingFailed,
+						};
+
+						self.payment_store.insert(payment)?;
```
Comment (jkczyz) [inline src/lib.rs:1351]: So we have any tests that check the payment store? Might be worth adding a test case for duplicate payments.

Comment (jkczyz) [reply src/lib.rs:1456]: But we don't check that the status remains the same as before in case of a duplicate error, IIUC.

**Review by jkczyz — COMMENTED** [review]
LGTM. Please squash.

**Review by jkczyz — APPROVED** [review]
